File: backend/services/stigmer-server/pkg/domain/skill/transfer/handler.go
package transfer

import (
	"errors"
	"fmt"
	"net/http"
	"strings"

	"github.com/rs/zerolog/log"

	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/domain/skill/storage"
)

// PathPrefix roots the transfer lane on the server's unified HTTP handler
// (the main gRPC port, which already routes /v1/proxy/* around gRPC-Web).
const PathPrefix = "/v1/skill-artifacts"

// uploadsSegment separates upload capabilities from download keys in the
// URL space: PUT {PathPrefix}/uploads/{ref} vs GET {PathPrefix}/{key}.
const uploadsSegment = "/uploads/"

// downloadKeyPrefix restricts the download surface to the skill artifact
// store's own keys ("skills/<hash>.zip"). Everything else under the storage
// root — should the two ever share one — stays unreachable from this lane.
const downloadKeyPrefix = "skills/"

// Sentinel errors let the HTTP handler map registry failures onto honest
// status codes without string matching.
var (
	errSlotUnknown  = errors.New("upload reference unknown or expired")
	errSlotConsumed = errors.New("upload reference already carries an upload")
	errSlotEmpty    = errors.New("upload reference has no uploaded bytes")
	errSizeMismatch = errors.New("upload size mismatch")
)

// UploadURL renders the capability URL for a minted reference against the
// lane's externally-reachable base URL. Kept next to the route definition so
// the URL shape and its handler cannot drift apart.
func UploadURL(baseURL, ref string) string {
	return fmt.Sprintf("%s%s%s%s", strings.TrimSuffix(baseURL, "/"), PathPrefix, uploadsSegment, ref)
}

// DownloadURL renders the download URL for an artifact storage key.
func DownloadURL(baseURL, storageKey string) string {
	return fmt.Sprintf("%s%s/%s", strings.TrimSuffix(baseURL, "/"), PathPrefix, storageKey)
}

// NewHandler serves the transfer lane:
//
//	PUT {PathPrefix}/uploads/{ref}  — stage artifact bytes (capability: ref)
//	GET {PathPrefix}/{storage_key}  — serve artifact bytes (capability: content-hash key)
//
// Neither route carries bearer auth by design: the URL is the credential,
// mirroring cloud's pre-signed R2 URLs. Minting an upload URL requires the
// same gRPC authorization as push; download keys are unguessable content
// hashes handed out by authorized skill reads.
func NewHandler(slots *UploadSlots, artifacts storage.ArtifactStorage) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		rest, ok := strings.CutPrefix(r.URL.Path, PathPrefix)
		if !ok {
			http.NotFound(w, r)
			return
		}

		if ref, isUpload := strings.CutPrefix(rest, uploadsSegment); isUpload {
			handleUpload(w, r, slots, ref)
			return
		}

		handleDownload(w, r, artifacts, strings.TrimPrefix(rest, "/"))
	})
}

func handleUpload(w http.ResponseWriter, r *http.Request, slots *UploadSlots, ref string) {
	if r.Method != http.MethodPut {
		w.Header().Set("Allow", http.MethodPut)
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	err := slots.Receive(ref, r.Body)
	switch {
	case err == nil:
		w.WriteHeader(http.StatusNoContent)
	case errors.Is(err, errSlotUnknown):
		// Deliberately the same shape for "never existed" and "expired":
		// distinguishing them would let an unauthorized caller probe which
		// tokens were once valid.
		http.Error(w, "upload reference unknown or expired — request a new upload URL", http.StatusNotFound)
	case errors.Is(err, errSlotConsumed):
		http.Error(w, "upload reference already used — request a new upload URL", http.StatusConflict)
	case errors.Is(err, errSizeMismatch):
		http.Error(w, err.Error()+" — the upload must match the size declared to createArtifactUploadUrl", http.StatusBadRequest)
	default:
		log.Error().Err(err).Str("ref", ref).Msg("skill artifact upload failed")
		http.Error(w, "failed to stage upload", http.StatusInternalServerError)
	}
}

func handleDownload(w http.ResponseWriter, r *http.Request, artifacts storage.ArtifactStorage, key string) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", http.MethodGet)
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	if !strings.HasPrefix(key, downloadKeyPrefix) {
		http.NotFound(w, r)
		return
	}

	// Full in-memory read matches the store's own interface (Get returns
	// bytes) and the ≤100MB validation ceiling bounds the allocation.
	data, err := artifacts.Get(key)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			http.NotFound(w, r)
			return
		}
		log.Error().Err(err).Str("storage_key", key).Msg("skill artifact download failed")
		http.Error(w, "failed to load artifact", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Length", fmt.Sprintf("%d", len(data)))
	w.Write(data)
}

File: backend/services/stigmer-server/src/domain/skill/transfer/handler.ts
/**
 * Skill artifact transfer lane — ports pkg/domain/skill/transfer/handler.go:
 * the HTTP upload/download surface that carries skill artifact bytes
 * OUTSIDE the gRPC control plane (#675). The gRPC server caps messages at
 * 10MB while the skill layer permits 100MB artifacts, so inline bytes
 * physically cannot carry every valid skill.
 *
 *   PUT {prefix}/uploads/{ref}  — stage artifact bytes (capability: ref)
 *   GET {prefix}/{storage_key}  — serve artifact bytes (capability: key)
 *
 * Neither route carries bearer auth by design: the URL is the credential,
 * mirroring cloud's pre-signed R2 URLs. Minting an upload URL requires the
 * same gRPC authorization as push; download keys are unguessable content
 * hashes handed out by authorized skill reads.
 *
 * The lane slots into the transport's existing skillTransferLane seam
 * (transport/server.ts lane 3); URL renderers live next to the route
 * dispatch so the URL shape and its handler cannot drift apart (Go's
 * stated invariant). Proven by __tests__/skill.test.ts's transfer-lane
 * block (both protocol stacks) and the conformance suite's transfer-lane
 * tests.
 */
import type { Logger } from "../../../boot/logger.js";
import { SKILL_ARTIFACTS_PATH_PREFIX } from "../../../transport/constants.js";
import type { LaneHandler, LaneRequest, LaneResponse } from "../../../transport/lanes.js";
import { DOWNLOAD_KEY_PREFIX, UPLOADS_SEGMENT } from "../constants.js";
import { ArtifactNotFoundError } from "../storage/artifact-storage.js";
import type { SkillArtifactStorage } from "../storage/artifact-storage.js";
import {
  SizeMismatchError,
  SlotConsumedError,
  SlotUnknownError,
} from "./slots.js";
import type { UploadSlots } from "./slots.js";

/**
 * Renders the capability URL for a minted reference against the lane's
 * externally-reachable base URL.
 */
export function uploadUrl(baseUrl: string, ref: string): string {
  return `${trimTrailingSlash(baseUrl)}${SKILL_ARTIFACTS_PATH_PREFIX}${UPLOADS_SEGMENT}${ref}`;
}

/** Renders the download URL for an artifact storage key. */
export function downloadUrl(baseUrl: string, storageKey: string): string {
  return `${trimTrailingSlash(baseUrl)}${SKILL_ARTIFACTS_PATH_PREFIX}/${storageKey}`;
}

function trimTrailingSlash(url: string): string {
  return url.endsWith("/") ? url.slice(0, -1) : url;
}

/** Builds the lane handler (Go NewHandler). */
export function newSkillTransferLane(
  slots: UploadSlots,
  artifacts: SkillArtifactStorage,
  logger: Logger,
): LaneHandler {
  return (request: LaneRequest, response: LaneResponse): void => {
    const url = request.url ?? "";
    const pathOnly = url.split("?")[0] ?? "";
    if (!pathOnly.startsWith(SKILL_ARTIFACTS_PATH_PREFIX)) {
      // Unreachable through the lane router (it dispatches by this exact
      // prefix); kept as the honest Go http.NotFound arm for direct use.
      writeText(response, 404, "404 page not found");
      return;
    }
    const rest = pathOnly.slice(SKILL_ARTIFACTS_PATH_PREFIX.length);

    // The lane router is synchronous; the async handlers run detached. The
    // terminal catch is load-bearing: a response-write failure on a
    // vanished client would otherwise escape as an unhandled rejection —
    // process-fatal under Node's default policy.
    const logDetachedFailure = (arm: string) => (error: unknown) => {
      logger.error(`skill transfer lane ${arm} handler failed`, {
        error: error instanceof Error ? error.message : String(error),
      });
    };

    if (rest.startsWith(UPLOADS_SEGMENT)) {
      const ref = rest.slice(UPLOADS_SEGMENT.length);
      handleUpload(request, response, slots, ref, logger).catch(
        logDetachedFailure("upload"),
      );
      return;
    }

    handleDownload(
      request,
      response,
      artifacts,
      rest.replace(/^\//, ""),
      logger,
    ).catch(logDetachedFailure("download"));
  };
}

async function handleUpload(
  request: LaneRequest,
  response: LaneResponse,
  slots: UploadSlots,
  ref: string,
  logger: Logger,
): Promise<void> {
  if (request.method !== "PUT") {
    response.setHeader("Allow", "PUT");
    writeText(response, 405, "method not allowed");
    return;
  }

  try {
    await slots.receive(ref, request);
    response.statusCode = 204;
    response.end();
  } catch (error) {
    if (error instanceof SlotUnknownError) {
      // Deliberately the same shape for "never existed" and "expired":
      // distinguishing them would let an unauthorized caller probe which
      // tokens were once valid.
      writeText(response, 404, "upload reference unknown or expired — request a new upload URL");
    } else if (error instanceof SlotConsumedError) {
      writeText(response, 409, "upload reference already used — request a new upload URL");
    } else if (error instanceof SizeMismatchError) {
      writeText(
        response,
        400,
        `${error.message} — the upload must match the size declared to createArtifactUploadUrl`,
      );
    } else {
      logger.error("skill artifact upload failed", {
        ref,
        error: error instanceof Error ? error.message : String(error),
      });
      writeText(response, 500, "failed to stage upload");
    }
  }
}

async function handleDownload(
  request: LaneRequest,
  response: LaneResponse,
  artifacts: SkillArtifactStorage,
  key: string,
  logger: Logger,
): Promise<void> {
  if (request.method !== "GET") {
    response.setHeader("Allow", "GET");
    writeText(response, 405, "method not allowed");
    return;
  }
  if (!key.startsWith(DOWNLOAD_KEY_PREFIX)) {
    writeText(response, 404, "404 page not found");
    return;
  }

  let data: Uint8Array;
  try {
    // Full in-memory read matches the store's own interface (get returns
    // bytes) and the ≤100MB validation ceiling bounds the allocation.
    data = await artifacts.get(key);
  } catch (error) {
    if (error instanceof ArtifactNotFoundError) {
      writeText(response, 404, "404 page not found");
      return;
    }
    logger.error("skill artifact download failed", {
      storageKey: key,
      error: error instanceof Error ? error.message : String(error),
    });
    writeText(response, 500, "failed to load artifact");
    return;
  }

  response.setHeader("Content-Type", "application/zip");
  response.setHeader("Content-Length", String(data.length));
  response.statusCode = 200;
  response.end(Buffer.from(data));
}

/** Go http.Error's shape: text/plain body with a trailing newline. */
function writeText(response: LaneResponse, status: number, body: string): void {
  response.statusCode = status;
  response.setHeader("Content-Type", "text/plain; charset=utf-8");
  response.end(`${body}\n`);
}
